### PR TITLE
Wayland windowing system

### PR DESCRIPTION
**Pull request by yol**
Add native Wayland windowing implementation as result of GSoC 2017 project from https://github.com/pkerling/xbmc/

## Description
This has been in the works for the last three months.

I have squashed the commits so that modifications and additions to the non-Wayland Kodi code are separate commits and that the Wayland windowing implementation itself is one commit. That implementation of course also has a history and is made up of a large number of commits (available via the linked GitHub repository) that were reviewed in feature chunks by @FernetMenta, but it would probably not be useful to keep it in mainline.

I realize that this is quite a lot of code to merge so I'm open to any suggestions that make it easier to handle.

Building with Wayland is fully integrated into the depends system and can be achieved by calling `tools/depends/configure` with `--enable-wayland`, so it should be easily integratable into Jenkins.

## Motivation and Context
Linux is transitioning to the Wayland protocol instead of X11 for the long term. To offer a competitive user experience on Linux and get all the shiny new features, Kodi must support Wayland natively instead of running via Xwayland X11 emulation.

## How Has This Been Tested?
Build and run test on linux64 with GL and GLES

## Screenshots (if appropriate):
It really looks just like on X :-)
(except for the window decorations which are better not shown off)

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
(This PR is mostly "new feature," but contains some commits that fix or improve core Kodi. I can list them individually if it helps.)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by fritsch** [comment]
Nice :-) I followed every issue and every issue review - really nice work. From my pov can go in every team if it does not break anything. Perhaps - only a suggest - it would be good to add "Wayland:" to the commits actually doing wayland features. This is especially useful when going just through a git log without seeing directly which file it changes.

But it's on your mentor to decide on this minor - only a suggestion.

**Comment by pkerling** [comment]
iOS is OK now, Windows fail is unrelated

Regarding actually building the Wayland stuff, should I just copy linux64 to linux64wayland so that it can be added to jenkins?

File: xbmc/cores/VideoPlayer/Process/wayland/ProcessInfoWayland.h [hunk 1]
```diff
@@ -1,7 +1,5 @@
-#pragma once
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Process/wayland/ProcessInfoWayland.h]: git moved the deleted file. I guess this is not what you wanted to achieve

File: xbmc/guilib/GraphicContext.cpp [hunk 2]
```diff
@@ -410,41 +411,138 @@ void CGraphicContext::SetVideoResolutionInternal(RESOLUTION res, bool forceUpdat
...
-  RESOLUTION_INFO info_org  = CDisplaySettings::GetInstance().GetResolutionInfo(res);
-
-  RESOLUTION_INFO info_mod = GetResInfo(res);
+  // FIXME Wayland windowing needs some way to "deny" resolution updates since what Kodi
+  // requests might not get actually set by the compositor.
+  // So in theory, m_iScreenWidth etc. would not need to be updated at all before the
+  // change is confirmed.
+  // But other windowing code expects these variables to be already set when
+  // SetFullScreen() is called, so set them anyway and remember the old values.
+  int origScreenWidth = m_iScreenWidth, origScreenHeight = m_iScreenHeight, origScreenId = m_iScreenId;
```
Comment (FernetMenta) [inline xbmc/guilib/GraphicContext.cpp]: put this on 3 separate lines

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.h [hunk 3]
```diff
@@ -196,6 +196,10 @@ class CRenderManager
   ERENDERSTATE m_renderState;
   CEvent m_stateEvent;
 
+  /// Display latency tweak value from AdvancedSettings for the current refresh rate
+  /// in seconds
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.h]: actually this is ms, right?

File: xbmc/settings/AdvancedSettings.cpp [hunk 4]
```diff
@@ -375,11 +375,7 @@ void CAdvancedSettings::Initialize()
 
   m_enableMultimediaKeys = false;
 
-#if defined(TARGET_DARWIN_IOS)
```
Comment (FernetMenta) [inline xbmc/settings/AdvancedSettings.cpp]: are you sure you want to change behavior for ios?

File: xbmc/utils/StringUtils.h [hunk 5]
```diff
@@ -143,6 +143,69 @@ class StringUtils
...
+
+   \param d_first the beginning of the destination range
+   \param input Input string to be split
+   \param delimiter Delimiter to be used to split the input string
+   \param iMaxStrings (optional) Maximum number of splitted strings
+   \return output iterator to the element in the destination range, one past the last element
+   *       that was put there
+   */
+  template<typename OutputIt>
+  static OutputIt SplitTo(OutputIt d_first, const std::string& input, const std::string& delimiter, unsigned int iMaxStrings = 0)
```
Comment (FernetMenta) [inline xbmc/utils/StringUtils.h]: IMO we should not have those fat bodies in header files.

File: xbmc/windowing/XBMC_events.h [hunk 6]
```diff
@@ -78,6 +80,10 @@ typedef struct XBMC_MoveEvent {
 	int y;		/* New y position */
 } XBMC_MoveEvent;
 
+struct XBMC_ModeChangeEvent {
```
Comment (FernetMenta) [inline xbmc/windowing/XBMC_events.h]: wrong format, we have brackets on next line

File: xbmc/windowing/XkbcommonKeymap.cpp [hunk 7]
```diff
@@ -0,0 +1,351 @@
+/*
```
Comment (FernetMenta) [inline xbmc/windowing/XkbcommonKeymap.cpp]: this file is only used by wayland. hence it should be placed into wayland folder

**Comment by FernetMenta** [comment]
Very good! Some minors then I think this is ready to go in.

File: xbmc/windowing/XkbcommonKeymap.cpp (hunk 7 above)
Comment (pkerling) [inline xbmc/windowing/XkbcommonKeymap.cpp]: It really should be used by mir, too, that's why I put it into a common folder. I can move the file or commit the mir change.

File: xbmc/utils/StringUtils.h (hunk 5 above)
Comment (pkerling) [inline xbmc/utils/StringUtils.h]: Having function bodies in headers is common C++ and for template functions there really is no alternative. Almost the whole STL is implemented like this.

File: xbmc/cores/VideoPlayer/Process/wayland/ProcessInfoWayland.h (hunk 1 above)
Comment (pkerling) [inline xbmc/cores/VideoPlayer/Process/wayland/ProcessInfoWayland.h]: The `Keymap.h` removal is a separate commit in fact. Git just detects this as move when producing the diff for the whole history. There's nothing you can do about it as far as I know.

File: xbmc/guilib/GraphicContext.cpp (hunk 2 above)
Comment (pkerling) [inline xbmc/guilib/GraphicContext.cpp]: done

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.h (hunk 3 above)
Comment (pkerling) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.h]: Yes, very good catch! Forgot to change this line during the unit change.

File: xbmc/settings/AdvancedSettings.cpp (hunk 4 above)
Comment (pkerling) [inline xbmc/settings/AdvancedSettings.cpp]: This does not really change iOS behavior. The only place this is used is in DisplaySettings (see below) where the check now includes `g_Windowing.CanDoWindowed()`, which is false for iOS. The only difference is that you now cannot force Kodi to go windowed anyway by modifying `advancedsettings.xml`. As this makes zero sense on iOS in either case I don't think it has any negative impact.

File: xbmc/windowing/XBMC_events.h (hunk 6 above)
Comment (pkerling) [inline xbmc/windowing/XBMC_events.h]: whole file is wrong format then, do you think it makes sense to only change this line?

File: xbmc/windowing/XkbcommonKeymap.cpp (hunk 7 above)
Comment (FernetMenta) [inline xbmc/windowing/XkbcommonKeymap.cpp]: don't bother with mir, maybe it dies soon. for now it should be moved to wayland.

File: xbmc/windowing/XBMC_events.h (hunk 6 above)
Comment (FernetMenta) [inline xbmc/windowing/XBMC_events.h]: sure it makes sense. the fact that someting is wrong is no justification to continue with doing it wrong.

File: xbmc/settings/AdvancedSettings.cpp (hunk 4 above)
Comment (FernetMenta) [inline xbmc/settings/AdvancedSettings.cpp]: ok

File: xbmc/windowing/XBMC_events.h (hunk 6 above)
Comment (pkerling) [inline xbmc/windowing/XBMC_events.h]: No I think it's maybe justification to reformat the whole file :-)

File: xbmc/utils/StringUtils.h (hunk 5 above)
Comment (FernetMenta) [inline xbmc/utils/StringUtils.h]: sure, but here we have a kind of ugly mixture. when I saw the method first, I did not see it was a template and could not find it in the body.
anyway, I have no strong feelings about this. If the team is ok with it, I am too

File: xbmc/windowing/XBMC_events.h (hunk 6 above)
Comment (FernetMenta) [inline xbmc/windowing/XBMC_events.h]: you are not the maintainer of this file. if you want to reformat it, you should to submit a separate pr for it. cosmetic changes just inflate the diff

Comment (pkerling) [inline xbmc/windowing/XBMC_events.h]: ok

File: xbmc/windowing/XkbcommonKeymap.cpp (hunk 7 above)
Comment (pkerling) [inline xbmc/windowing/XkbcommonKeymap.cpp]: ok

**Comment by pkerling** [comment]
OK I think that was all of the minors

File: tools/depends/native/scons-native/Makefile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+$(TARBALLS_LOCATION)/$(ARCHIVE):
+	cd $(TARBALLS_LOCATION); $(RETRIEVE_TOOL) $(RETRIEVE_TOOL_FLAGS) $(BASE_URL)/$(ARCHIVE)
+
+$(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
+	-rm -rf $(PLATFORM)/*; mkdir -p $(PLATFORM)
+	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
+
+.installed-$(PLATFORM): $(PLATFORM)
+	cd $(PLATFORM); $(PREFIX)/bin/python setup.py install --prefix=$(PREFIX)
```
Comment (Rechi) [inline tools/depends/native/scons-native/Makefile]: does scons not have a build step?

File: tools/depends/target/wayland-protocols/Makefile [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+CONFIGURE=./configure --prefix=$(PREFIX)
+
+CLEAN_FILES=$(ARCHIVE) $(PLATFORM)
+
+all: .installed-$(PLATFORM)
+
+$(TARBALLS_LOCATION)/$(ARCHIVE):
+	cd $(TARBALLS_LOCATION); $(RETRIEVE_TOOL) $(RETRIEVE_TOOL_FLAGS) $(BASE_URL)/$(ARCHIVE)
+
+$(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
```
Comment (Rechi) [inline tools/depends/target/wayland-protocols/Makefile]: something like https://github.com/xbmc/xbmc/blob/94faa07/tools/depends/target/crossguid/Makefile#L48-L52 is missing

File: tools/depends/target/waylandpp/Makefile [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+SCONS=scons
+export CROSSCXX:=$(CXX)
+export CXX=$(CXX_FOR_BUILD)
+
+all: .installed-$(PLATFORM)
+
+$(TARBALLS_LOCATION)/$(ARCHIVE):
+	cd $(TARBALLS_LOCATION); $(RETRIEVE_TOOL) $(RETRIEVE_TOOL_FLAGS) $(BASE_URL)/$(ARCHIVE)
+
+$(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
```
Comment (Rechi) [inline tools/depends/target/waylandpp/Makefile]: something like https://github.com/xbmc/xbmc/blob/94faa07/tools/depends/target/crossguid/Makefile#L48-L52 is missing

File: tools/depends/configure.ac [hunk 11]
```diff
@@ -89,6 +89,12 @@ AC_ARG_ENABLE([gplv3],
   [use_gplv3=$enableval],
   [use_gplv3=yes])
 
+AC_ARG_ENABLE([wayland],
+  [AS_HELP_STRING([--enable-wayland],
+  [build libraries needed for Wayland. default is no])],
+  [enable_wayland=$enableval],
+  [enable_wayland=no])
```
Comment (Rechi) [inline tools/depends/configure.ac]: IMO there should be a check if wayland can be used for the target platform.
OSX, ios, and android can't use it

File: tools/depends/native/scons-native/Makefile (hunk 8 above)
Comment (pkerling) [inline tools/depends/native/scons-native/Makefile]: It's python setuptools, so it's debatable. I think you *can* call `setup.py build`, but not sure how useful that would be. tbh I just c&p-ed from https://github.com/xbmc/xbmc/blob/master/tools/depends/native/distribute-native/Makefile

File: tools/depends/target/wayland-protocols/Makefile (hunk 9 above)
Comment (pkerling) [inline tools/depends/target/wayland-protocols/Makefile]: added

File: tools/depends/target/waylandpp/Makefile (hunk 10 above)
Comment (pkerling) [inline tools/depends/target/waylandpp/Makefile]: added

File: tools/depends/configure.ac (hunk 11 above)
Comment (pkerling) [inline tools/depends/configure.ac]: added

**Comment by pkerling** [comment]
Just give me a ping if this is ready for squashing

**Comment by Rechi** [comment]
please switch to the new style of component logging which was introduced at #12676

**Comment by pkerling** [comment]
> please switch to the new style of component logging which was introduced at #12676

Done. Had to rebase anyway, squashed while at it.

**Review by FernetMenta — APPROVED** [review]

**Comment by fritsch** [comment]
@FernetMenta  when do you push the button? Need to buy champagne before that I think :-)

**Comment by FernetMenta** [comment]
if no objections, I will merge this this evening.

File: cmake/platform/linux/wayland.cmake [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+set(PLATFORM_OPTIONAL_DEPS VAAPI)
+
+set(WAYLAND_RENDER_SYSTEM "" CACHE STRING "Render system to use with Wayland: \"gl\" or \"gles\"")
+
+if(WAYLAND_RENDER_SYSTEM STREQUAL "gl")
+  list(APPEND PLATFORM_REQUIRED_DEPS OpenGl)
+elseif(WAYLAND_RENDER_SYSTEM STREQUAL "gles")
+  list(APPEND PLATFORM_REQUIRED_DEPS OpenGLES)
+else()
+  message(SEND_ERROR "You need to decide whether you want to use GL- or GLES-based rendering in combination with the Wayland windowing system. Please set WAYLAND_RENDER_SYSTEM to either \"gl\" or \"gles\". For normal desktop systems, you will usually want to use \"gl\".")
```
Comment (wsnipex) [inline cmake/platform/linux/wayland.cmake]: Can we set sane defaults here? I.e gl on x86 and gles on arm

Comment (pkerling) [inline cmake/platform/linux/wayland.cmake]: @FernetMenta was against that and he convinced me. I also think now that it's better to not have a hardly predictable default that depends on CPU architecture and/or installed libraries.

Comment (wsnipex) [inline cmake/platform/linux/wayland.cmake]: I am pretty sure this will cause confusion. Having gl as default on x86 is hardly unpredictable

Comment (FernetMenta) [inline cmake/platform/linux/wayland.cmake]: it is, most wayland application most likely use GLES. this has absolutely nothing to do with processor arch.

Comment (pkerling) [inline cmake/platform/linux/wayland.cmake]: > I am pretty sure this will cause confusion. Having gl as default on x86 is hardly unpredictable

Hm, not sure why it would be confusing to have to choose. The message also explicitly states to try `gl` on normal systems (we can of course reword the message if it isn't clear enough). And if you're compiling for embedded you should know whether you want GLES or not.

Comment (wsnipex) [inline cmake/platform/linux/wayland.cmake]: How about making it dependent on which libs have been found/enabled? I.e if GL was enabled by the user, use them

Comment (FernetMenta) [inline cmake/platform/linux/wayland.cmake]: If someone builds Kodi, she/he should know what to build for and why. Otherwise refrain from building. Those kind of defaults are evil, you end up with an executable that may not fit your expectations.

Comment (wsnipex) [inline cmake/platform/linux/wayland.cmake]: In case it wasn't clear. We already have options for selecting gl or gles, those should be used instead of inventing another option that does the same thing.

Comment (pkerling) [inline cmake/platform/linux/wayland.cmake]: > We already have options for selecting gl or gles, those should be used instead of inventing another option that does the same thing.

I don't think there really is a strong precedent like you are implying. The only platform that at the moment even allows building with GL *or* GLES is mir, which is not used extensively. All other platforms are fixed to either GL or GLES.

Comment (pkerling) [inline cmake/platform/linux/wayland.cmake]: And the problem with the options we have (if you mean ENABLE_... cmake variables) is the following: Where should I add GL and GLES to in `wayland.cmake`? To `PLATFORM_REQUIRED_DEPS` or `PLATFORM_OPTIONAL_DEPS`? Neither fits in my opinion if you leave the decision up to ENABLE_...

Comment (wsnipex) [inline cmake/platform/linux/wayland.cmake]: I'd add them to optional, so both get searched for, but still retain the check here if one was enabled

Comment (pkerling) [inline cmake/platform/linux/wayland.cmake]: I can do that of course, but I think it's confusing because they are not in fact optional. You'd also have to disallow setting either of them to `AUTO` (since configuring must fail if it's not found). The current system isn't quite cut out for more complex dependencies like "you need either X or Y" :-/

Comment (wsnipex) [inline cmake/platform/linux/wayland.cmake]: True, those are valid concerns. Another option would be ENABLE_WAYLAND=OPENGL/OPENGLES

Comment (pkerling) [inline cmake/platform/linux/wayland.cmake]: I'm not sure I understand. You want to rename the variable from `WAYLAND_RENDER_SYSTEM` to `ENABLE_WAYLAND`?

Comment (wsnipex) [inline cmake/platform/linux/wayland.cmake]: I'm sorry, hard to keep an overview on a phone during holidays. I forgot that Wayland is enabled with core_platform name and not via enable_Wayland. So this doesn't make sense. Let s just keep what you have for now and see how it goes. We can always optimize later.

**Comment by FernetMenta** [comment]
congratualtions and many thanks to @pkerling

**Comment by MartijnKaijser** [comment]
Congrats and well done. Also for the other semi related fixes.

**Comment by fritsch** [comment]
No need to stop helping and improving kodi for us! Thanks very much for this milestone work that pushes kodi to the next level into a great standard linux future.

**Comment by paulpach** [comment]
This was exciting to watch.  This was the only reason why I don't have fedora in my HTPC,  A million thanks to @pkerling

**Comment by da-anda** [comment]
congrats and thanks a lot pkerling

**Comment by pkerling** [comment]
And there it goes 🎉  Thanks a lot to @FernetMenta for reviewing all of this stuff and giving valuable advice at every stage :-)

**Comment by raneon** [comment]
Thank you so much pkerling!!! Highly appreciated :-)
